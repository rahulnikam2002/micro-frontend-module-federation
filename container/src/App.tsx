import React from 'react';
import './App.css';
const Header = React.lazy(() => import('remote/Header'));


function App() {
  return (
    <div className="App">
      <React.Suspense fallback={<h1>Loading Header...</h1>}>
        <Header />
      </React.Suspense>
    </div>
  );
}

export default App;
