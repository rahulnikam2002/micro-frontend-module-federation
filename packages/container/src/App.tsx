import React from 'react';
import './App.css';
const Header = React.lazy(() => import('remote/Header'));
const Footer = React.lazy(() => import('footer/Footer'));
const GlobalButton = React.lazy(() => import('footer/GlobalButton'));


function App() {
  return (
    <div className="App">
      <React.Suspense fallback={<h1>Loading Header...</h1>}>
        <Header />
      </React.Suspense>

      <React.Suspense fallback={<h1>Loading Button...</h1>}>
        <GlobalButton />
      </React.Suspense>

      <React.Suspense fallback={<h1>Loading Footer...</h1>}>
        <Footer />
      </React.Suspense>
    </div>
  );
}

export default App;
