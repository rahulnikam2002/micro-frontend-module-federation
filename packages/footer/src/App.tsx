import React from 'react';
import './App.css';

function App() {
  return (
    <div className="App">
      <h1>Upcoming micro frontend - Footer</h1>
    </div>
  );
}

export default App;
