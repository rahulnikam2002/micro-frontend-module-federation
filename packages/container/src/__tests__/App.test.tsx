// App.test.tsx
import React from 'react';
import "@testing-library/jest-dom";
import { render, screen } from '@testing-library/react';

// Mock the remotes for Jest
jest.mock('remote/Header', () => ({
  __esModule: true,
  default: () => <div data-testid="mock-header">Mock Header</div>,
}));

jest.mock('footer/Footer', () => ({
  __esModule: true,
  default: () => <div data-testid="mock-footer">Mock Footer</div>,
}));

// Mock GlobalButton used by App
jest.mock(
  'footer/GlobalButton',
  () => ({
    __esModule: true,
    default: () => <button data-testid="mock-global-button">Mock Button</button>,
  }),
  { virtual: true }
);

import App from '../App'; // Import AFTER mocking

describe('App', () => {
  test('renders lazy-loaded Header and Footer with Suspense', async () => {
    render(<App />);

    // Check Suspense fallback
    expect(screen.getByText(/loading header/i)).toBeInTheDocument();
    expect(screen.getByText(/loading footer/i)).toBeInTheDocument();
    expect(screen.getByText(/loading button/i)).toBeInTheDocument();

    // Wait for lazy-loaded components
    const header = await screen.findByTestId('mock-header');
    const footer = await screen.findByTestId('mock-footer');
    const button = await screen.findByTestId('mock-global-button');

    expect(header).toBeInTheDocument();
    expect(footer).toBeInTheDocument();
    expect(button).toBeInTheDocument();
  });

  test('renders components in the correct order and matches snapshot', async () => {
    const { container } = render(<App />);

    // Wait for lazy-loaded components to resolve
    const header = await screen.findByTestId('mock-header');
    const footer = await screen.findByTestId('mock-footer');
    const button = await screen.findByTestId('mock-global-button');

    // Ensure the order inside .App is Header, Button, Footer
    const appDiv = container.querySelector('.App');
    expect(appDiv).toBeInTheDocument();
    const children = Array.from(appDiv!.children).filter(
      (n) => n.nodeType === Node.ELEMENT_NODE
    );

  // Find indexes (use the actual nodes rendered inside .App)
  const idxHeader = children.indexOf(header as Element);
  const idxButton = children.indexOf(button as Element);
  const idxFooter = children.indexOf(footer as Element);

    // The header should come before the button, and button before footer
    expect(idxHeader).toBeGreaterThanOrEqual(0);
    expect(idxButton).toBeGreaterThanOrEqual(0);
    expect(idxFooter).toBeGreaterThanOrEqual(0);
    expect(idxHeader).toBeLessThan(idxButton);
    expect(idxButton).toBeLessThan(idxFooter);

    // Snapshot the rendered output for extra coverage
    expect(appDiv).toMatchSnapshot();
  });
});
