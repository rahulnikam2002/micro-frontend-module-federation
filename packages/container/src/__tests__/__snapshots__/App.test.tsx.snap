// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`App renders components in the correct order and matches snapshot 1`] = `
<div
  class="App"
>
  <div
    data-testid="mock-header"
  >
    Mock Header
  </div>
  <button
    data-testid="mock-global-button"
  >
    Mock Button
  </button>
  <div
    data-testid="mock-footer"
  >
    Mock Footer
  </div>
</div>
`;
